sycl/include/CL/sycl/types.hpp

```diff
@@ -2274,6 +2274,73 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
 #undef __SYCL_DECLARE_SCALAR_BOOL_CONVERTER
 #undef __SYCL_USE_EXT_VECTOR_TYPE__
 
+/// is_device_copyable is a user specializable class template to indicate
+/// that a type T is device copyable. If is_device_copyable is specialized such
+/// that is_device_copyable_v<T> == true on a T that does not satisfy all
+/// the requirements of a device copyable type, the results are unspecified.
+template <typename T, typename = void>
+struct is_device_copyable : std::false_type {};
+
+template <typename T>
+struct is_device_copyable<
+    T, std::enable_if_t<std::is_trivially_copyable<T>::value>>
+    : std::true_type {};
+
+#if __cplusplus >= 201703L
+template <typename T>
+inline constexpr bool is_device_copyable_v = is_device_copyable<T>::value;
+#endif // __cplusplus >= 201703L
+
+// std::tuple<> is implicitly device copyable type.
+template <> struct is_device_copyable<std::tuple<>> : std::true_type {};
+
+// std::tuple<Ts...> is implicitly device copyable type if each type T of Ts...
+// is device copyable.
+template <typename T, typename... Ts>
+struct is_device_copyable<std::tuple<T, Ts...>>
+    : detail::bool_constant<is_device_copyable<T>::value &&
+                            is_device_copyable<std::tuple<Ts...>>::value> {};
+
+namespace detail {
+template <typename T, typename = void>
+struct IsDeprecatedDeviceCopyable : std::false_type {};
+
+// TODO: using C++ attrubute [[deprecated]] or the macro __SYCL2020_DEPRECATED
+// does not produce expected warning message for the type 'T'.
+template <typename T>
+struct __SYCL2020_DEPRECATED("This type isn't device copyable in SYCL 2020")
+    IsDeprecatedDeviceCopyable<
+        T, std::enable_if_t<std::is_trivially_copy_constructible<T>::value &&
+                            std::is_trivially_destructible<T>::value &&
+                            !is_device_copyable<T>::value>> : std::true_type {};
+
+#ifdef __SYCL_DEVICE_ONLY__
+template <typename T, unsigned NumToCheck> struct CheckDeviceCopyableHelper;
+
+template <typename T> struct CheckDeviceCopyableHelper<T, 0> {};
+
+template <typename T> struct CheckDeviceCopyableHelper<T, 1> {
+  using FieldT = decltype(__builtin_field_type(T, 0));
+  static_assert(is_device_copyable<FieldT>::value ||
+                    detail::IsDeprecatedDeviceCopyable<FieldT>::value,
+                "Not device copyable!");
+};
+
+template <typename T, unsigned NumToCheck>
+struct CheckDeviceCopyableHelper
+    : CheckDeviceCopyableHelper<T, NumToCheck - 1> {
+  using FieldT = decltype(__builtin_field_type(T, NumToCheck - 1));
+  static_assert(is_device_copyable<FieldT>::value ||
+                    detail::IsDeprecatedDeviceCopyable<FieldT>::value,
+                "Not device copyable!");
+};
+
+template <typename FuncT>
+struct CheckDeviceCopyable
+    : CheckDeviceCopyableHelper<FuncT, __builtin_num_fields(FuncT)> {};
```

Comment:
FuncT here is either a) lambda function type or b) a functor type with corresponding operator() accepting id<Dims> or nd_item<Dims> parameter. Those function types cannot have base fields, right?

Reply:
Can you please help make a test case with a FunctorT passes to parallel_for() such that that (__builtin_num_bases(FunctorT) > 0) ? The test case above seems unrelated to checks of objects captured by lambda/functor, or I do not understand something.

Reply:
how about:
````
struct Base { /* stuff that makes this non-device copyable*/
};

struct FunctorType : Base {
int i;
void operator()() const{}
};
...
kernel_single_task<>(FunctorType{}); // Should be rejected, because of 'Base', but this code will NOT reject it, since it is only checking the fields, not the bases.


````

Reply:
> 
> 
> FuncT here is either a) lambda function type or b) a functor type with corresponding operator() accepting id or nd_item parameter. Those function types cannot have base fields, right?

Functor types can ABSOLUTELY have a base type.  Lambdas cannot, but Functor types are simply classes with an operator().

Reply:
We want to diagnose code like this, don't we?
```
class KernelFunctor : public std::map<std::string, volatile int> {
public:
  void operator()(item<1> it) const;
};
```

Reply:
I agree and I don't say functors cannot have base classes. What I am talking is the functor's operator()(item<Dims>), or operator()().
For kernel it only matters what operator() captures, i.e. the test case below seems legal to me and it is compiled successfully (without this PR changes): 

`#include <CL/sycl.hpp>
#include <string>
#include <map>
using namespace cl::sycl;

struct KernelFunctor : public std::map<std::string, volatile int> {
  void operator()() const {};
};

int main() {
  queue Q;
  Q.submit([&](handler &CGH) {
    CGH.single_task< class XYZ >(KernelFunctor());
  });
  return 0;
}
`

Reply:
At least based on the implementation we did before, that is incorrect. Also, that example SHOULD have been caught, I'm surprised our visitor didn't get it the last time...

There is no such thing as a 'capture' in an operator(), nor a field.  The reason we check fields and bases is because we have to copy the entire object (which is what we do with lambdas as well, except the fields to that structure are auto-generated).

Reply:
Also, that exact example fails: http://cmplrexplorer.intel.com/z/4jrPn5

Reply:
> Also, that exact example fails: http://cmplrexplorer.intel.com/z/4jrPn5

Interesting, I used a github.com/intel/llvm built on Windows and it silently passed, while Linux compiler failed.
In any case, I see the idea now and will fix the implementation by using __builtin_num_bases.

Reply:
b24fab2650813f4790c50649d8446eac90eb494c checks both: fields & bases. Thank you!